### PR TITLE
Post-editor reduxification: add redux actions #1

### PR DESCRIPTION
**Pull request by artpi**
# What?

Part of https://github.com/Automattic/wp-calypso/issues/2993

Introduces redux actions side-by-side flux actions

This is the first step in Editor reduxification

1. Duplicate redux actions
2. See what do they do when they are together
3. Work on reducers, action creators
4. Hook state back into these components
5. Phase out flux and libs

# This PR
This introduces a portion of actions and a groundwork.
Whole process can be split into several PRs, but this one has to be merged to lay the groundwork (actions file)

# Further work
Progress tracking in #2993

### CODE REVIEW

File: client/post-editor/edit-post-status/index.jsx [hunk 1]
```diff
@@ -54,6 +61,7 @@ var EditPostStatus = React.createClass( {
 		stats.recordEvent( 'Changed Sticky Setting', stickyEventLabel );
 
 		actions.edit( { sticky: ! this.props.post.sticky } );
+		this.props.toggleStickyStatus( this.props.post.sticky );
```
Comment (artpi) [inline client/post-editor/edit-post-status/index.jsx]: I purposefully copied the `toggleStickyStatus `, I think the whole logic of upper-scope `toggleStickyStatus ` function should be moved to action.
I'm even considering moving the logic to `actions.js` right now.
@mtias , what do you thing

Comment (artpi) [inline client/post-editor/edit-post-status/index.jsx]: Summary:
```
do we want individual actions or a general edit action? etc

Artur Piszek [1:22 PM] 
Individual

Matías Ventura [1:22 PM] 
I think so too

Artur Piszek [1:22 PM] 
I'm advocating for individual as an abstraction

Matías Ventura [1:22 PM] 
specially since we may have a component outside of the editor that does just that one action

[1:23] 
but it could also complicate things a lot since a lot of those actions are connected

Artur Piszek [1:23 PM] 
And we already have `toggleStickyStatus`

[1:23] 
In component

[1:24] 
So I think in the end it should migrate from compoennt to actions

Matías Ventura [1:24 PM] 
I agree with that

Artur Piszek [1:24 PM] 
In actions.js they can use general edit action if we want

[1:24] 
But we want to be able to import just a single action

[1:24] 
Ok, so I will continue in that fashion
```

Comment (mtias) [inline client/post-editor/edit-post-status/index.jsx]: Makes sense, but let's do that cleanup after we get all the actions in place. It would help us evaluate what we have to deal with better.

File: client/state/ui/editor/post/actions.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+export function startEditingNew( site, postOptions ) {
+	debug( 'startEditingNew', site, postOptions );
+	return { type: 'TODO' };
+}
+
+export function startEditingExisting( site, postID ) {
+	debug( 'startEditingExisting', site, postID );
+	return { type: 'TODO' };
```
Comment (gwwar) [inline client/state/ui/editor/post/actions.js]: Note that the original fires 'START_EDITING_POST' and also performs an ajax call, where 'RECEIVE_POST_TO_EDIT' fires on successful callback.

I know the you're mirroring the original name, but I wonder if we could improve on this later to something like `loadPostAndTrackEdits`

File: client/post-editor/edit-post-status/index.jsx [hunk 3]
```diff
@@ -53,7 +60,9 @@ var EditPostStatus = React.createClass( {
 		stats.recordStat( stickyStat );
 		stats.recordEvent( 'Changed Sticky Setting', stickyEventLabel );
 
+		// TODO: REDUX - remove flux actions when whole post-editor is reduxified
```
Comment (gwwar) [inline client/post-editor/edit-post-status/index.jsx]: Maybe toggle it above, so it's clearer.
```
const isSticky = ! this.props.post.sticky;
actions.edit( { sticky: isSticky } );
this.props.toggleStickyStatus( isSticky );
```

File: client/post-editor/editor-author/index.jsx [hunk 4]
```diff
@@ -15,6 +15,10 @@ var Gravatar = require( 'components/gravatar' ),
 	config = require( 'config' ),
 	stats = require( 'lib/posts/stats' );
 
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
```
Comment (aduth) [inline client/post-editor/editor-author/index.jsx]: These two imports should be kept with `External dependencies`

File: client/post-editor/editor-author/index.jsx [hunk 5]
```diff
@@ -15,6 +15,10 @@ var Gravatar = require( 'components/gravatar' ),
...
+import { changeAuthor } from 'state/ui/editor/post/actions';
+
 var EditorAuthor = React.createClass( {
```
Comment (aduth) [inline client/post-editor/editor-author/index.jsx]: By introducing `import` syntax, this is now an ESLint warning. Consider using `const` instead of `var`.

File: client/post-editor/editor-categories/index.jsx [hunk 6]
```diff
@@ -17,7 +17,11 @@ var CategorySelector = require( 'my-sites/category-selector' ),
 	siteUtils = require( 'lib/site/utils' ),
 	stats = require( 'lib/posts/stats' );
 
-module.exports = React.createClass( {
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
```
Comment (aduth) [inline client/post-editor/editor-categories/index.jsx]: These two imports should be kept with `External dependencies`

File: client/post-editor/editor-delete-post/index.jsx [hunk 7]
```diff
@@ -13,7 +13,11 @@ import utils from 'lib/posts/utils';
 import Gridicon from 'components/gridicon';
 import Tooltip from 'components/tooltip';
 
-export default React.createClass( {
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
```
Comment (aduth) [inline client/post-editor/editor-delete-post/index.jsx]: These two imports should be kept with `External dependencies`

File: client/post-editor/editor-discussion/index.jsx [hunk 8]
```diff
@@ -13,6 +13,10 @@ var EditorFieldset = require( 'post-editor/editor-fieldset' ),
 	InfoPopover = require( 'components/info-popover' ),
 	stats = require( 'lib/posts/stats' );
 
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
```
Comment (aduth) [inline client/post-editor/editor-discussion/index.jsx]: These two imports should be kept with `External dependencies`

File: client/post-editor/editor-drawer/index.jsx [hunk 9]
```diff
@@ -37,6 +37,10 @@ var Accordion = require( 'components/accordion' ),
 	userSettings = require( 'lib/user-settings' ),
 	AppPromo = require( 'components/app-promo' );
 
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
```
Comment (aduth) [inline client/post-editor/editor-drawer/index.jsx]: These two imports should be kept with `External dependencies`

File: client/post-editor/editor-featured-image/index.jsx [hunk 10]
```diff
@@ -1,3 +1,10 @@
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
```
Comment (aduth) [inline client/post-editor/editor-featured-image/index.jsx]: These two imports should be kept with `External dependencies`

File: client/post-editor/editor-featured-image/index.jsx [hunk 11]
```diff
@@ -1,3 +1,10 @@
...
+import {
+	setFeaturedImage,
+	removeFeaturedImage
+} from 'state/ui/editor/post/actions';
```
Comment (aduth) [inline client/post-editor/editor-featured-image/index.jsx]: These imports should be kept with `Internal dependencies`

File: client/state/ui/editor/post/actions.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import debugFactory from 'debug';
```
Comment (aduth) [inline client/state/ui/editor/post/actions.js]: Should include `External dependencies` docblock.

File: client/state/ui/editor/post/actions.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+	debug( 'setFeaturedImage', newImage );
+	return { type: 'TODO' };
+}
+
+export function removeFeaturedImage() {
+	debug( 'removeFeaturedImage' );
+	return { type: 'TODO' };
+}
+
+export const EDITING_MODE_EXISTING = 'EXISTING';
```
Comment (aduth) [inline client/state/ui/editor/post/actions.js]: All action types should be defined in `state/action-types.js` ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#actions)).

File: client/post-editor/editor-author/index.jsx [hunk 14]
```diff
@@ -55,7 +59,9 @@ var EditorAuthor = React.createClass( {
 	onSelect: function( author ) {
 		stats.recordStat( 'advanced_author_changed' );
 		stats.recordEvent( 'Changed Author' );
+		// TODO: REDUX - remove flux actions when whole post-editor is reduxified
 		PostActions.edit( { author: author } );
+		this.props.changeAuthor( author );
```
Comment (gwwar) [inline client/post-editor/editor-author/index.jsx]: Do you prefer change over set?

File: client/state/ui/editor/post/actions.js (hunk 13 above)
Comment (aduth) [inline client/state/ui/editor/post/actions.js]: Action type value should match the variable name.

Comment (aduth) [inline client/state/ui/editor/post/actions.js]: Oh... these aren't action types? Hmm...

File: client/post-editor/editor-discussion/test/index.jsx [hunk 15]
```diff
@@ -48,7 +48,7 @@ describe( 'EditorDiscussion', function() {
 			recordEvent: noop,
 			recordStat: noop
 		} );
-		EditorDiscussion = require( '../' );
+		EditorDiscussion = require( '../' ).WrappedComponent;
```
Comment (aduth) [inline client/post-editor/editor-discussion/test/index.jsx]: Why are we importing the wrapped component? If the connected component is not ready to use, should we not be changing it at all here?

**Comment by aduth** [comment]
>This introduces a portion of actions and a groundwork.

The introduction of actions seems a bit scattered. I wonder if we should be taking a different approach here, by starting to implement individual actions before attempting to wire them to existing components. Additionally, I think we should be more intentional about which actions we're converting, likely one per pull request.

File: client/post-editor/test/post-editor.jsx [hunk 16]
```diff
@@ -56,6 +56,10 @@ mockery.registerMock( 'my-sites/drafts/draft-list', MOCK_COMPONENT );
 mockery.registerMock( 'lib/layout-focus', {
 	set() {}
 } );
+// TODO: REDUX - add proper tests when whole post-editor is reduxified
```
Comment (gwwar) [inline client/post-editor/test/post-editor.jsx]: We can add tests where we spy that the expected redux function has been called in certain editor actions. There's no need to wait for the action to be fully implemented. If there isn't much scaffolding for editor unit tests, this may be a bit of work.

File: client/state/ui/editor/post/actions.js (hunk 13 above)
Comment (gwwar) [inline client/state/ui/editor/post/actions.js]: Maybe create a const object to avoid confusion:
```javascript
const EDITING_MODES = { EXISTING: 'EXISTING', NEW: 'NEW' };
```

File: client/post-editor/test/post-editor.jsx (hunk 16 above)
Comment (artpi) [inline client/post-editor/test/post-editor.jsx]: Well, thats true, but following tests are actually rendering the component with all its children.
This needs further deliberation, because I kind of wouldnt call them unit tests.
In any case, this is the test of whole post editor and in my opnion, rewriting it should wait until post editor redux moe is finished.

**Comment by artpi** [comment]
> The introduction of actions seems a bit scattered

Absolutely. I'm going in alphabetical order.

> I think we should be more intentional about which actions we're converting, likely one per pull request.

This PR is introducing 12 of them out of 40 I have found to this point (possibly more). If there's going to be 1 per pull request, then it's going to take forever.

> I wonder if we should be taking a different approach here, by starting to implement individual actions

I'm totally opposed to that.

We dont even know what actions do we have. I intent to stub them all in `editor/post/actions.js` and then we will have overlook and can decide how to do it properly.
If we start by implementing individual actions, we will have perfectly designed actions that we have to rewrite components to use.
We can do that, but we'll make more informed decisions if we have these actions in one file.

In this PR I'm not deciding on any shape of actions, just connecting components and listing existing actions.

File: client/post-editor/edit-post-status/index.jsx (hunk 3 above)
Comment (artpi) [inline client/post-editor/edit-post-status/index.jsx]: I totally agree.
But in an effort to keep this PR mergeable, I'm not touching previous behaviour right now.
In my opinion this whole code should be moved to redux action and analytics handled by middleware

File: client/state/ui/editor/post/actions.js (hunk 13 above)
Comment (artpi) [inline client/state/ui/editor/post/actions.js]: Yeah, you're both right :)
I went with @gwwar suggestion: 99064bbbb557065de611d103636a34dfbcfc14c5

File: client/state/ui/editor/post/actions.js (hunk 2 above)
Comment (artpi) [inline client/state/ui/editor/post/actions.js]: Yes, I agree.
I desparately try to avoid any decisions in this PR series.
We should definitely ditch callbacks, but I intend these PRs to provide us framework to think about this stuff.

File: client/post-editor/editor-author/index.jsx (hunk 14 above)
Comment (artpi) [inline client/post-editor/editor-author/index.jsx]: Good point. Changed in 0fd337c3bbe19b3db82c40daa58b9e8e1e6936fe

File: client/post-editor/editor-author/index.jsx (hunk 4 above)
Comment (artpi) [inline client/post-editor/editor-author/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

File: client/post-editor/editor-author/index.jsx (hunk 5 above)
Comment (artpi) [inline client/post-editor/editor-author/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

File: client/post-editor/editor-categories/index.jsx (hunk 6 above)
Comment (artpi) [inline client/post-editor/editor-categories/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

File: client/post-editor/editor-delete-post/index.jsx (hunk 7 above)
Comment (artpi) [inline client/post-editor/editor-delete-post/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

File: client/post-editor/editor-discussion/index.jsx (hunk 8 above)
Comment (artpi) [inline client/post-editor/editor-discussion/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

File: client/post-editor/editor-drawer/index.jsx (hunk 9 above)
Comment (artpi) [inline client/post-editor/editor-drawer/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

File: client/post-editor/editor-featured-image/index.jsx (hunk 10 above)
Comment (artpi) [inline client/post-editor/editor-featured-image/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

File: client/post-editor/editor-featured-image/index.jsx (hunk 11 above)
Comment (artpi) [inline client/post-editor/editor-featured-image/index.jsx]: Good point! Fixed in e02fb9d1a2a71b379c4bac097aa3d18c8331660b

**Comment by artpi** [comment]
Thank you @gwwar and @aduth for taking the time :) I have adressed all of the comments

File: client/state/ui/editor/post/actions.js [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+	debug( 'startEditingNew', site, postOptions );
+	return { type: 'TODO' };
+}
+
+export function startEditingExisting( site, postID ) {
+	debug( 'startEditingExisting', site, postID );
+	return { type: 'TODO' };
+}
+
+export function toggleStickyStatus( currentStatus ) {
```
Comment (aduth) [inline client/state/ui/editor/post/actions.js]: Is there some discussion to why these actions have been split out to individual properties, than our [existing approach](https://github.com/Automattic/wp-calypso/blob/3c8fb4f98ecc713370ebbb9feaa5883321f60b33/client/lib/posts/actions.js#L237-L247) where we accept an object of updated attributes?

Comment (artpi) [inline client/state/ui/editor/post/actions.js]: As a matter of fact, there was with @mtias : https://github.com/Automattic/wp-calypso/pull/3025#discussion_r51764414

Arguments pro-individual-property:
- Most of [this](https://github.com/Automattic/wp-calypso/pull/3025/files#diff-96fbb5e8d8b2d3c5308a18e3b9b20161R59), [this](https://github.com/Automattic/wp-calypso/pull/3025/files#diff-c65fc9514aaa064f1fa8671d6b735e28R51) and similar function logic should be ported from component to actions and reducers. Using proper abstraction allows us to do so in the future.
- In the actions file we can underneath use the old approach, [like in notices](https://github.com/Automattic/wp-calypso/blob/master/client/state/notices/actions.js#L43)
- Component to be logically pure should import only the action it needs instead of one mega-setter
- It allows us to restructure underlying data object only by manipulating those setters
- We have to have actionCreators either way. Having individual ones has no downside

File: client/state/ui/editor/post/actions.js (hunk 12 above)
Comment (artpi) [inline client/state/ui/editor/post/actions.js]: Done: dd69cf3cdc13134faca94c87a04305b68f87f725

**Comment by artpi** [comment]
@gwwar always suggests adding `PropTypes`, so I added them and defaultprops in an preemptive manner.

**Comment by aduth** [comment]
I don't think I'm not doing very well in articulating myself, but I'm at odds with the idea of duplicating all of the existing actions without first considering at least some rough idea of how we want the Redux editor state structure to look (#2994). There are a number of poor choices that were made in implementing the Flux actions that we should reconsider, and fundamental differences in how we'll want to track editor state in Redux. In our Flux stores, we had no notion of tracking multiple post drafts, or posts themselves for that matter. In approaching this problem by first copying assumptions made in the Flux implementation, I worry that we'll not have a solid foundation upon which the editor can be backed by a Redux-managed state. We can promise to evaluate and refactor in subsequent pull requests, but in my experience, we (generally) don't do well at holding ourselves responsible for refactoring, and I suspect these patterns will persist indefinitely.

This is not to say that we should try to implement the state behaviors independent of any integration with the editor, which I probably mistakenly conveyed in an earlier comment. Instead, I think we can convert individual actions used by the editor in isolation, with careful consideration of how it fits with our current data recommendations. For example, here are a few top-level problems that we can tackle now in small, focused pull requests, that likely differ from our current Flux-based implementation:

- How does the editor become aware of the saved copy of a post? In our Flux implementation, this was implemented as part of `PostActions.startEditingExisting`, which triggered a network request and dispatched actions asynchronously. In our Redux state, I think we ought not tie post data to the editor so tightly, since posts are used broadly across the application. Instead, we could start by implementing a `requestSitePost( siteId, postId )` action in the `posts` Redux subtree, then move on to implementing a `<QueryPost />` query component to be rendered by the editor.
- For that matter, do we need to distinguish `startEditingExisting` from `startEditingNew` at all? Since this was only needed previously in that the former action would trigger a network request, we could probably remove this distinction altogether, simplifying to a single `startEditing` action creator. We can start by implementing this action creator, the reducer for managing the state, and duplicating existing behavior by adding the new action dispatching to the editor route controller.

In the end, both approaches will have duplicated existing behavior in managing editor state before starting to convert individual components to consuming exclusively from the Redux state tree, but what I'm suggesting is more up-front planning in addressing what are fundamental differences between Flux and Redux state implementations.

_TLDR: Flux !== Redux_

I'm also still opposed to individual property action creators, but I feel I may already be agitating far too much already :smile:

**Comment by artpi** [comment]
@aduth: I agree with most of this. But this is also an approach that will be VERY LONG to implement and we need post-editor reduxified ASAP (I mean like 2-3 weeks).
Having this kind of conversation before actually implementing has proven to postpone any action for a long time.
The approach I'm taking here is basically listing all the action so that we have information to answer questions you have asked.
Nothing is set in stone, this is an interim state. I agree that these tend to live very long, but honestly we dont know what we dont know.

> first copying assumptions made in the Flux implementation

Well, you have to see these assuptions in one place to make any informed decision.

> Instead, I think we can convert individual actions used by the editor in isolation, 

Well, this approach will mean that we'll be stuck in exactly what you are worried about: We will be stuck in this precise action and it will be very hard to combine, split or refactor this action because you are - in isolation. I intent to squish it into one place so we can see what can we merge, refactor and so on.

It will also take forever to do it in isolation.

I really DO share your concerns and it hurts me that I am copying the old approach. But if we want to have any hope or moving forward, we need to battle-test our approaches.
Part of individual property action creators is that they are easy to find and replace in codebase.

When I'm finished with placing stubs and BEFORE implementing action creator bodies we can do exactly what you advocate - have a look what we need and mix & match.
But it is impossible to do this operation without any concrete examples.

@mtias @rralian - I need you to weigh in on this before I can continue or drop it

File: client/state/ui/editor/post/actions.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
+/**
```
Comment (mtias) [inline client/state/ui/editor/post/actions.js]: I don't think we should put this under UI. I think several (most?) of these actions should be in the `posts` sub-tree. Anything else that is intrinsically specific to editor could be in a new `state.editor` portion.

**Comment by rralian** [comment]
> @mtias @rralian - I need you to weigh in on this before I can continue or drop it

I think this is a reasonable approach. @aduth is worried that if we start by copying all of the actions that we won't give any thought to whether they make sense as is or can be improved. I understand the concern, but I think if you look at this as an in-progress feature, we're not really putting anything in stone. I tend to agree with @artpi's point:
> you have to see these assuptions in one place to make any informed decision.

and
> I desparately try to avoid any decisions in this PR series.
We should definitely ditch callbacks, but I intend these PRs to provide us framework to think about this stuff.

So I would suggest, let's finish this up, get the rest of the actions in place in followup PRs, and then _before we start implementing the reducers_, do a hangout (our next calypso-core one maybe) to discuss pieces that we want to change. That can include modifying the actions. @aduth and @mtias have the most experience with the editor data flow and any of the pieces they'd like to move around, so they can help provide that direction. But I know I personally would find it much easier to reason about this when we have everything listed out in one place.

File: client/state/ui/editor/post/actions.js (hunk 18 above)
Comment (rralian) [inline client/state/ui/editor/post/actions.js]: Do we want to discuss that here or in https://github.com/Automattic/wp-calypso/issues/2994?

Comment (artpi) [inline client/state/ui/editor/post/actions.js]: @mtias @rralian : This was also my concern.
I think we need to keep it 1 file in order to achieve "overview" of all post editor actions.
We already have `state/ui/editor/media` and `state/ui/editor/contactform` so I think this makes sense for now.
Notice that I am not implementing any reducers nor selectors right now. And I think actions can be split up in the future.
But I will reiterate that we need to gather them all in 1 place before we start to shove them in proper places.

**Comment by artpi** [comment]
I suggest continuing in a following way:

1. Continue this PRs and copy actions to one file (`state/editor/post/actions.js`)
2. Once its in one file, browse through the file and consider following:
    - splitting actions to proper places in the redux tree
    - bundling actions together
    - reshaping actions
    - reshaping redux tree
    - moving logic from components to actions
    - changing the actions logic
3. Host a hangout to discuss all of this
4. Prepare proper redux tree shape
5. Start refactoring and implementing in isolated PRs

**Comment by aduth** [comment]
Separating actions contradicts the goal of "not deciding on any shape of actions" as it is an opinionated change, but I relent and agree that we should move forward with duplication and revisit in future pull requests. :+1:
